### DISCUSSION TITLE
extension util method for FXGL?

### DISCUSSION THREAD

**Discussion by chengenzhao** [AlmasB/FXGL#1035] [Ideas]
Hi:

Is it possible to add in two more extra functions for key down&up events?
There are two lambda parameters, 1st for keydown event, 2nd key up event
and could bind more than 1 key codes into these runnable code

```java
  public static void onKey(Runnable keyDownFunc, Runnable keyUpFunc, KeyCode... keyCodes){
    for(var keyCode:keyCodes){
      FXGL.getInput().addAction(new UserAction(keyCode.name()) {
        @Override
        protected void onActionBegin() {
          keyDownFunc.run();
        }

        @Override
        protected void onActionEnd() {
          keyUpFunc.run();
        }
      }, keyCode);
    }

  }

  public static void onKey(Runnable keyDownFunc, KeyCode... keyCodes){
    onKey(keyDownFunc,()->{},keyCodes);
  }
```

and we use it like
```java
    onKey(() -> player.getComponent(Player.class).moveRight(),
      () -> player.getComponent(Player.class).stopMovingRight(), KeyCode.D, KeyCode.RIGHT);

    onKey(() -> player.getComponent(Player.class).moveLeft(),
      () -> player.getComponent(Player.class).stopMovingLeft(), KeyCode.A, KeyCode.LEFT);
```

**Comment by AlmasB** [comment]
The fluent API for input building supports this: 

https://github.com/AlmasB/FXGL/blob/50679d556cc46b732b09bb2822ded45faea8ad45/fxgl-samples/src/main/java/intermediate/InputFluentSample.java#L23-L30

**Reply by chengenzhao** [reply]
and support multiple keys?
like this:
```java
onKey(()-> player.getComponent(Player.class).fire(),
      () -> player.getComponent(Player.class).stopFire(), 
      KeyCode.J,KeyCode.K,KeyCode.L,KeyCode.U,KeyCode.I,KeyCode.O,KeyCode.Y,KeyCode.H,KeyCode.N,KeyCode.M);
```

**Reply by AlmasB** [reply]
Whilst the above can be repeated for multiple keys if needed, I do not wish to encourage users to bind the same action to multiple triggers of the same type. Mainly because, the same action would be fired multiple times, which then forces users to handle this behaviour themselves. For example:

```
player.getComponent(Player.class).fire()
```

will be called for each key specified above in the frame these keys were pressed, allowing the player to fire multiple times in the same frame.

**Reply by chengenzhao** [reply]
Yeah
this will simplify players' operations
they don't need to read specifications or instructions to find the shooting key, they just press any keys and the player just fire ammos.

**Reply by AlmasB** [reply]
If this behaviour is intended, I would just do:

```
List.of(KeyCode.A, KeyCode.B, ...).forEach(key -> {
    onKeyBuilder(key) 
             .onActionBegin(() -> System.out.println("Action Begin")) 
             .onAction(() -> System.out.println("On Action")) 
             .onActionEnd(() -> System.out.println("Action End")); 
});
```
